### ISSUE TITLE
Task detection errors

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#47016]
While verifying https://github.com/Microsoft/vscode/issues/45816

- Disable all task autodetction

```
    "typescript.tsc.autoDetect": "off",
    "jake.autoDetect": "off",
    "gulp.autoDetect": "off",
    "npm.autoDetect": "off",
    "grunt.autoDetect": "off"
```

- Type `task ` in quickopen.
- Get lots of errors like `Error: The gulp task detection didn't contribute a task for the following configuration: ...` and I can't select those tasks

Maybe I just misunderstand what "autodetect" is. I would assume it just powers the detection of tasks in package.json, gulpfile.js, etc, that aren't configured in a tasks.json. But if I try to execute a configured task of type "gulp", I should still be able to run it using the gulp command. 

If that's not right, then what is the point of disabling autodetect?

**Comment by dbaeumer** [comment]
I agree that this can be improved but current tasks of type gulp, npm, ... (the auto detected onces) can't be defined in a tasks.json file. They can only be configured (tweaked) and must be detected. Otherwise the task runner doesn't know how to call gulp. This is only know by the extension contributing the gulp task. I will keep the issue open since this can be improved with the resolve API a task provider can implement.

**Comment by gonssal** [comment]
I have a project with a "npm run watch" task. This is a task that should be run only once, but everytime I want to save a sass file I got asked which task to run, and when I select the "watch" one, I still get asked more things. This happens everytime I save the file, no matter if the task is already running or not.

To avoid this really annoying behavior I tried setting `"npm.autoDetect": "off"`. Now everytime I want to save the file, I get asked which task to run and the only option is "No build task to run found. Configure Build Task...".

The only way to save the file is click "Ctrl + S" twice, then pressing "Escape" to close the tasks dialog. This has made working with SASS files in a project with a related npm task unusable and I had to switch to another editor.

Just wanted to explain that in my opinion, this is a showstopper right now, at least for me.

**Comment by dbaeumer** [comment]
@gonssal I am pretty sure that this is caused by an extension. Can you please try to run code with `--disable-extensions` from a terminal and see if this still happens. If not please find the extension and report the issue there.

**Comment by gonssal** [comment]
@dbaeumer Just ran with --disable-extensions, exactly the same behavior. Even tried restarting and even logging out after adding `"npm.autoDetect": "off"`.

**Comment by dbaeumer** [comment]
@gonssal can you provide me with a GitHub repository I can clone that demos the behavior ?

**Comment by dbaeumer** [comment]
Any steps for that repository. Usually I would prefer a smaller test case :-(

**Comment by gonssal** [comment]
Clone the repo, edit `resources/assets/sass/app.scss`, see the behavior happen. Then add `"npm.autoDetect": "off"` to the settings, see the behavior still happening with no available tasks.

**Comment by dbaeumer** [comment]
I can't reproduce this using the provided repository. I am still 99% convinced that this is caused by an extension since I am not aware of any code in VS Code that auto executes tasks. Please note that when you start code with --disable-extensions you need to close all VS Code windows beforehand. Here is what I get when making edits to the file:

![cast](https://user-images.githubusercontent.com/1931590/38980461-314ff8aa-43bd-11e8-97f3-b01c78cf2d0a.gif)

**Comment by alpaix** [comment]
@dbaeumer Do you have an estimate on when this issue can be resolved?

This bug appears to be very annoying in Live Share scenarios. It occurs when a guest joins a collaboration session. We modify the guest workspace settings and disable all task autodetection in favor of Live Share task provider that fetches workspace tasks from the host. So every time the guest user opens `Run Task...` menu the error in Tasks output channel is shown and even worse the Tasks output channel steals the focus
```
Error: The npm task detection didn't contribute a task for the following configuration:
{
    "type": "npm",
    "script": "start",
    "problemMatcher": [
        "$eslint-compact"
    ],
    "isBackground": true
}
The task will be ignored.
```

**Comment by dbaeumer** [comment]
Even if the resolve API is in it will not resolve your problem. What you actually want is to take over the whole task execution including parsing the `tasks.json` file. There could be a custom task provider for which you don't know any disablement setting. It couldn't even have one. In addition the task engine would if a gulp task exists in the tasks.json still ask the gulp extension to resolve it if the user selects it and then execute it. But if I understand you correctly you want to not have this happening at all. Right?

If this is the case an easier trick would be to not surface the tasks.json file at all on the guest side by hiding it in your file system provider. The auto detection code for gulp, ... will do nothing even if they are enabled since the workspace folder URI is not a `file` uri and hence the folder will be ignored. Does that make sense?

**Comment by dbaeumer** [comment]
Even better would be to surface the tasks.json under a different name / path. The the guest can still edit it but the tasks don't show up locally. Only the once you mirror.

**Comment by thomas-darling** [comment]
So here's the thing...

It's very common to have a `package.json` defining a few primary scripts that you may run, and several child-scripts, that are only intended to be run by one of the primary scripts.

It would be really handy if we could disable task auto-detection, and configure only the primary scripts as tasks in `tasks.json`. That way, we won't see all the child scripts.

Right now, my tasks list is a complete and utter mess because of this, so I'd really appreciate a solution to it. Maybe have an option to, instead of completely disabling task detection, just use the `tasks.json` file as a filter, and show only the configured tasks?

Oh, and yeah - that error message about task contribution needs to be improved. I really expected this to work, and was quite surprised by that strange error - I mean, I thought I had explicitly told it to not auto-detect, because I'd rather configure the tasks myself...

**Comment by alexr00** [comment]
The original issue is solved now because `resolveTask` is implemented. If there are other problems please feel free to ping this issue, or, even better, open a new issue with the details.
